schema-templates/basic-auth.sql

```diff
@@ -0,0 +1,298 @@
+-------------------------------------------------------------------------------
+-- Adapted from https://github.com/robconery/pg-auth
+
+begin;
+
+create extension if not exists pgcrypto;
+create schema if not exists basic_auth;
+
+-------------------------------------------------------------------------------
+-- Utility functions
+
+create or replace function
+basic_auth.random_value(len int, out result varchar(32)) as
+$$
+BEGIN
+SELECT substring(md5(random()::text),0, len) into result;
+END
+$$ LANGUAGE plpgsql;
+
+
+create or replace function
+basic_auth.clearance_for_role(u name) returns void as
+$$
+declare
+  ok boolean;
+begin
+  select exists (
+    select rolname
+      from pg_authid
+     where pg_has_role(current_user, oid, 'member')
+       and rolname = u
+  ) into ok;
+  if not ok then
+    raise invalid_password using message =
+      'current user not member of role ' || u;
+  end if;
+end
+$$ LANGUAGE plpgsql;
+
+-------------------------------------------------------------------------------
+-- Login storage and constraints
+
+create table if not exists
+basic_auth.logins (
+  username character varying not null,
+  pass   character(60) not null,
+  role   name not null,
+  email  character varying not null unique,
+  active boolean not null default false,
+  more   JSON,
+  constraint l_pkey primary key (username)
+);
+
+create or replace function
+basic_auth.check_role_exists() returns trigger
+  language plpgsql
+  as $$
+begin
+  if not exists (select 1 from pg_roles as r where r.rolname = new.role) then
+    raise foreign_key_violation using message =
+      'unknown database role: ' || new.role;
+    return null;
+  end if;
+  return new;
+end
+$$;
+
+drop trigger if exists ensure_login_role_exists on basic_auth.logins;
+create constraint trigger ensure_login_role_exists
+  after insert or update on basic_auth.logins
+  for each row
+  execute procedure basic_auth.check_role_exists();
+
+create or replace function
+basic_auth.encrypt_pass() returns trigger
+  language plpgsql
+  as $$
+begin
+  if tg_op = 'INSERT' or new.pass <> old.pass then
```

Comment:
I actually raised the limit to 512 and added limits to all fields in this public table so people can't dump huge data in it.